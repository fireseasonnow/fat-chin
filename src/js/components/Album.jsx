// @flow
import React from 'react';
import Track from './Track';
import styled from 'styled-components'

type Props = {
    index: number,
    name: string,
    coverUrl: string,
    year: number,
    tracks: Array<string>,
    tracksVisible: boolean,
    tracksVisibilityHandler: (e: Object, index: number) => void,
    coverZoomHandler: (e: Object, index: number) => void,
    showLyricsHandler: (e: Object, name: string) => void
};

const album = (props: Props) => {
    const { name, coverUrl, year, tracks, tracksVisible, index, tracksVisibilityHandler, coverZoomHandler, showLyricsHandler } = props;
    const tracksVisiblityText = tracksVisible ? 'Hide tracks' : 'Show tracks';

    return (
        <Item>
            <Title>{name} ({year})</Title>
            <CoverWrapper>
                <CoverImg src={coverUrl} alt={`'${name}' album cover`} />
                <Overlay>
                    <OverlayItem onClick={(e) => tracksVisibilityHandler(e, index)} type="button">{tracksVisiblityText}</OverlayItem>
                    <OverlayItem onClick={(e) => coverZoomHandler(e, index)} type="button">Zoom cover</OverlayItem>
                </Overlay>
            </CoverWrapper>

            <Tracks>
                {tracksVisible ? tracks.map((track, index) => (
                    <Track
                        key={parseInt(index, 10)}
                        number={parseInt(index) + 1}
                        name={track}
                        showLyricsHandler={showLyricsHandler}>
                    </Track>
                )) : null}
            </Tracks>
        </Item>
    );
};

export default album;

const Item = styled.li`
    list-style: none;
`;

const Title = styled.h2`
    margin: 0;
    font-size: 1em;
    text-align: center;
`;

const CoverImg = styled.img`
    display: block;
    max-width: 100%;
    height: auto;
    filter: grayscale(1);
    transition: filter ease-in-out 250ms;

    ${Item}:hover & {
        filter: grayscale(0);
        transition: filter ease-in-out 250ms;
    }
`;

const CoverWrapper = styled.div`
    position: relative;
`;

const Tracks = styled.ul`
    padding-left: 0;
`;

const Overlay = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    position: absolute;
    bottom: 0;
    width: 100%;
    height: 50px;
    background: rgba(255, 255, 255, 0.7);
    opacity: 0;
    transition: opacity ease-in-out 250ms;

    ${CoverWrapper}:hover & {
        opacity: 1;
        transition: opacity ease-in-out 250ms;
    }
`;

const OverlayItem = styled.button`
    display: flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    color: #000;
    background: rgba(255, 255, 255, 0.05);
    cursor: pointer;
    font-family: 'Oswald', sans-serif;
    font-size: 1em;
    outline: none;
    border: 0;

    &:first-child,
    &:first-child:active {
        border-right: 1px solid #999;
    }
`;
