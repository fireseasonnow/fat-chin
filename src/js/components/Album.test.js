import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';

import Album from './Album';

configure({ adapter: new Adapter() });

describe('<Album />', () => {
    let wrapper = {};

    beforeEach(() => {
        wrapper = shallow(<Album />);
    });

    test('should render without crashing', () => {
        expect(wrapper).toBeTruthy();
    });

    test('should render an album without tracks and \'Show tracks\' text', () => {
        wrapper.setProps({
            name: 'Hello',
            coverUrl: 'url',
            year: 1984,
            tracksVisible: false,
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    test('should render an album with tracks and \'Hide tracks\' text', () => {
        wrapper.setProps({
            name: 'Hello',
            coverUrl: 'url',
            year: 1984,
            tracksVisible: true,
            tracks: [
                'blah',
                'blah-blah'
            ],
        });

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    test('should fire tracksVisibilityHandler when clicked on a tracks button', () => {
        const tracksVisibilityHandler = jest.fn();

        wrapper.setProps({
            index: 4,
            tracksVisibilityHandler,
        });

        const event = { target: { value: 'test' } };
        wrapper.find('.qa-album__overlay-item--tracks-btn').simulate('click', event.target.value, 4);

        expect(tracksVisibilityHandler).toHaveBeenCalledWith(event.target.value, 4);
    });

    test('should fire coverZoomHandler when clicked on a zoom button', () => {
        const coverZoomHandler = jest.fn();

        wrapper.setProps({
            index: 3,
            coverZoomHandler,
        });

        const event = { target: { value: 'test' } };
        wrapper.find('.qa-album__overlay-item--zoom-btn').simulate('click', event.target.value, 3);

        expect(coverZoomHandler).toHaveBeenCalledWith(event.target.value, 3);
    });
});
