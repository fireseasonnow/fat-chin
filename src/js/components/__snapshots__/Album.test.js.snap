// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Album /> should render an album with tracks and 'Hide tracks' text 1`] = `
<li
  className="album__item"
>
  <h1
    className="album__name"
  >
    Hello
     (
    1984
    )
  </h1>
  <div
    className="album__cover-wrapper"
  >
    <img
      alt="'Hello' album cover"
      className="album__img"
      src="url"
    />
    <div
      className="album__overlay"
    >
      <button
        className="album__overlay-item qa-album__overlay-item--tracks-btn"
        onClick={[Function]}
        type="button"
      >
        Hide tracks
      </button>
      <button
        className="album__overlay-item qa-album__overlay-item--zoom-btn"
        onClick={[Function]}
        type="button"
      >
        Zoom cover
      </button>
    </div>
  </div>
  <ul
    className="album__tracks"
  >
    <track
      key="0"
      name="blah"
      number={1}
    />
    <track
      key="1"
      name="blah-blah"
      number={2}
    />
  </ul>
</li>
`;

exports[`<Album /> should render an album without tracks and 'Show tracks' text 1`] = `
<li
  className="album__item"
>
  <h1
    className="album__name"
  >
    Hello
     (
    1984
    )
  </h1>
  <div
    className="album__cover-wrapper"
  >
    <img
      alt="'Hello' album cover"
      className="album__img"
      src="url"
    />
    <div
      className="album__overlay"
    >
      <button
        className="album__overlay-item qa-album__overlay-item--tracks-btn"
        onClick={[Function]}
        type="button"
      >
        Show tracks
      </button>
      <button
        className="album__overlay-item qa-album__overlay-item--zoom-btn"
        onClick={[Function]}
        type="button"
      >
        Zoom cover
      </button>
    </div>
  </div>
  <ul
    className="album__tracks"
  />
</li>
`;
